File: 0907-koko-eating-bananas/0907-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int max=Integer.MIN_VALUE;
        for(int curr: piles){
            max = Math.max(max, curr);
        }
        int left=1,right=max;
        int min_hrs=Integer.MAX_VALUE;
        while(left<=right){
            int mid=left+(right-left)/2;
            if(computeHours(piles,mid)<=h){
                right=mid-1;
                min_hrs=Math.min(min_hrs,mid);
            }else{
                left=mid+1;
            }
        }
        return min_hrs;
    }
    public int computeHours(int[] piles, int eat_rate){
        int total=0;
        for(int i:piles){
            total+=Math.ceil((double)i/(double)eat_rate);
        }
        return total;
    }
}

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # first we have to find a high value of k, so we can have a range to search; so we take the max of the piles
        low = 1
        high = max(piles)
        while low < high:
            mid = (low + high) // 2
            if not self.checkIfKokoIsAbleToFinish(piles, mid, h):
               low= mid+1
            else:
                high = mid
        return low

    def checkIfKokoIsAbleToFinish(self, piles, eating_speed, max_hours):
        time = 0
        for pile in piles:
            time += math.ceil(pile / eating_speed)
            if time > max_hours:
                return False
        return True
